### PR TITLE
Fixed transaction status for relayed v1 & v2 intra-shard transactions

### PR DESCRIPTION
**Pull request by iulianpascalau**
- fixed transaction status for relayed v1 & v2 intra-shard transactions

### CODE REVIEW

File: process/testdata/finishedOKRelayedTxIntraShard.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    ],
+    "operation": "transfer",
+    "initiallyPaidFee": "1152000000000000",
+    "fee": "1152000000000000",
+    "isRelayed": true,
+    "chainID": "D",
+    "version": 1,
+    "options": 0
+  }
+}
```
Comment (sstanculeanu) [inline process/testdata/finishedOKRelayedTxIntraShard.json:46]: missing new line

File: process/testdata/finishedFailedRelayedTxIntraShard.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    ],
+    "operation": "transfer",
+    "initiallyPaidFee": "1052500000000000",
+    "fee": "1052500000000000",
+    "isRelayed": true,
+    "chainID": "T",
+    "version": 1,
+    "options": 0
+  }
+}
```
Comment (sstanculeanu) [inline process/testdata/finishedFailedRelayedTxIntraShard.json:54]: missing new line

File: process/transactionProcessor.go [hunk 3]
```diff
@@ -421,6 +422,16 @@ func (tp *TransactionProcessor) computeTransactionStatus(tx *transaction.ApiTran
 		return data.TxStatusUnknown
 	}
 
+	// sanity checks
+	senderAddress, err := tp.pubKeyConverter.Decode(tx.Sender)
+	if err != nil {
+		return transaction.TxStatusFail
```
Comment (andreibancioiu) [inline process/transactionProcessor.go:428]: If sender is _**metachain**_ (e.g. reward transactions), works as expected?

https://testnet-gateway.multiversx.com/transaction/191a0bf2d29559181da9d93bfd35f8b9f136122006d10b51096fec6be65fb25d

On failure to decode, perhaps `TxStatusUnknown`?

Comment (andreibancioiu) [reply process/transactionProcessor.go:428]: It seems that for reward transactions, `process-status` currently works fine:

https://testnet-gateway.multiversx.com/transaction/191a0bf2d29559181da9d93bfd35f8b9f136122006d10b51096fec6be65fb25d/process-status

File: process/transactionProcessor.go [hunk 4]
```diff
@@ -484,6 +500,26 @@ func checkIfMoveBalanceNotarized(tx *transaction.ApiTransactionResult) bool {
 	return true
 }
 
+func (tp *TransactionProcessor) checkIfIntraShardRelayedTransaction(
```
Comment (andreibancioiu) [inline process/transactionProcessor.go:503]: This checks relayer shard & user shard, but not shard of `innerTx.receiver`. If inner tx is cross-shard, but relayed one (wrapper) is intra-shard, a "false positive" success might happen (even if tx is still pending). If so, then the check maybe should be "is tx relayed & completely intra-shard, all actors in a shard".

I am not sure about this, though, so - it's just a question, not a comment.

File: process/transactionProcessor.go (hunk 3 above)
Comment (iulianpascalau) [reply process/transactionProcessor.go:428]: whoa, thanks for the input. Will add these examples and will come up with a fix

File: process/transactionProcessor.go (hunk 4 above)
Comment (iulianpascalau) [reply process/transactionProcessor.go:503]: yup, agree, will come up with a fix for this :)

File: process/testdata/finishedOKRelayedTxIntraShard.json (hunk 1 above)
Comment (iulianpascalau) [reply process/testdata/finishedOKRelayedTxIntraShard.json:46]: done

File: process/testdata/finishedFailedRelayedTxIntraShard.json (hunk 2 above)
Comment (iulianpascalau) [reply process/testdata/finishedFailedRelayedTxIntraShard.json:54]: done

File: process/transactionProcessor.go (hunk 3 above)
Comment (iulianpascalau) [reply process/transactionProcessor.go:428]: done

File: process/transactionProcessor.go [hunk 5]
```diff
@@ -500,24 +500,48 @@ func checkIfMoveBalanceNotarized(tx *transaction.ApiTransactionResult) bool {
...
+		return false, err
+	}
+	receiverAddress, err := tp.pubKeyConverter.Decode(tx.Receiver)
+	if err != nil {
+		return false, err
+	}
+
+	dataField := string(tx.Data)
+	if strings.Index(dataField, relayedTxV1DataMarker) != 0 {
+		return false, fmt.Errorf("wrong relayed v1 data marker position")
```
Comment (andreibancioiu) [inline process/transactionProcessor.go:525]: Correct, `ProcessingTypeOn*` identifies Relayed V1, and the check is for Relayed V1 :ballot_box_with_check:

**Review by sstanculeanu — DISMISSED** [review]

File: process/transactionProcessor.go [hunk 6]
```diff
@@ -500,24 +500,48 @@ func checkIfMoveBalanceNotarized(tx *transaction.ApiTransactionResult) bool {
...
+	}
+
+	innerTx := &transaction.Transaction{}
+	err = tp.relayedTxsMarshaller.Unmarshal(innerTx, innerTxData)
+	if err != nil {
+		return false, err
 	}
 
-	isSameShard := tp.proc.GetShardCoordinator().SameShard(senderAddress, receiverAddress)
+	isSameShardOnRelayed := tp.proc.GetShardCoordinator().SameShard(senderAddress, receiverAddress)
```
Comment (andreibancioiu) [inline process/transactionProcessor.go:540]: Can also be written as 2 checks / 2 calls of `SameShard()`: `(senderWhichIsRelayer, innerTx.SndAddr)` and `(innerTx.SndAddr, innerTx.RcvAddr)` - but all right to stay as it is, as well, it's more explicit.

Comment (andreibancioiu) [reply process/transactionProcessor.go:540]: About relayed V1 that wrap NFT transfers: not-yet-completed transactions of such an exotic kind might reach the "is relayed completely intrashard" check, and can be mistaken as completely intrashard, and thus reach an earlier "done" status - even though they are pending.

Even so, this is an extremely exotic case (no occurrence on mainnet). I think we can skip it for now, what do you think?

**Review by andreibancioiu — APPROVED** [review]

**Review by cristure — DISMISSED** [review]

File: process/transactionProcessor.go (hunk 6 above)
Comment (iulianpascalau) [reply process/transactionProcessor.go:592]: refactored the solution

File: process/transactionProcessor.go [hunk 7]
```diff
@@ -500,7 +499,50 @@ func checkIfMoveBalanceNotarized(tx *transaction.ApiTransactionResult) bool {
 	return true
 }
 
-func (tp *TransactionProcessor) checkIfCompletelyIntraShardRelayedTransaction(tx *transaction.ApiTransactionResult) (bool, error) {
+func (tp *TransactionProcessor) addMissingLogsOnProcessingExceptions(
```
Comment (andreibancioiu) [inline process/transactionProcessor.go:502]: :+1:

**Review by andreibancioiu — APPROVED** [review]

**Review by cristure — DISMISSED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Review by cristure — APPROVED** [review]

**Review by andreibancioiu — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]

**Pull request merged**